src/notation/internal/notationparts.cpp

```diff
@@ -706,7 +706,7 @@ void NotationParts::replaceInstrument(const ID& instrumentId, const ID& fromPart
         return;
     }
 
-    part->setInstrument(InstrumentsConverter::convertInstrument(newInstrument), oldInstrumentInfo.fraction);
+    score()->undo(new Ms::ChangePart(part, new Ms::Instrument(InstrumentsConverter::convertInstrument(newInstrument)), part->name()));
     doSetPartName(part, formatPartName(part));
```

Comment:
It does concern me that I'm now discarding oldInstrumentInfo.fraction, but I can't find a case that it's anything other than -1 - in fact that seems to be a bug, because in MU3 if you use "staff properties" at a point after a "Change Instrument" event, it shows the transposition etc. info as set by that "change instrument" event and that's what you can edit - whereas for MU4 "Change Instrument" events seem to be ignored in the staff properties dialog.
MU3 has a bunch of code in editstaff.cpp (lines 381-) to support this, but the MU4 code doesn't pass any info about the current score position when doing this.